pkg/finality-grandpa/voter.go

```diff
@@ -0,0 +1,1114 @@
+
+		if maybeCompletedLastRound != nil {
+			lastRound := *maybeCompletedLastRound
+			lastRoundState = *lastRound.bridgeState()
+			pastRounds.Push(env, lastRound)
+		}
+
+		// when there is no information about the last completed round,
+		// the best we can do is assume that the estimate == the given base
+		// and that it is finalized. This is always the case for the genesis
+		// round of a set.
+	}
+
+	bestRound := newVotingRound(
+		lastRoundNumber+1,
+		voters,
+		lastFinalized,
+		&lastRoundState,
+		finalizedSender,
+		env,
+	)
+
+	inner := &innerVoterState[Hash, Number, Signature, ID, Environment[Hash, Number, Signature, ID]]{
+		bestRound:  bestRound,
+		pastRounds: *pastRounds,
+	}
+	globalOut := make(chan CommunicationOut)
+	return &Voter[Hash, Number, Signature, ID]{
+		env:                    env,
+		voters:                 voters,
+		inner:                  inner,
+		finalizedNotifications: finalizedNotifications,
+		lastFinalizedNumber:    lastFinalizedNumber,
+		lastFinalizedInRounds:  lastFinalized,
+		globalIn:               globalIn,
+		globalOut:              newBuffered(globalOut),
+		stopChan:               make(chan any),
+	}, globalOut
+}
+
+func (v *Voter[Hash, Number, Signature, ID]) pruneBackgroundRounds(waker *waker) error {
+	v.inner.Lock()
+	defer v.inner.Unlock()
+
+pastRounds:
+	for {
+		// Do work on all background rounds, broadcasting any commits generated.
+		ready, nc, err := v.inner.pastRounds.pollNext(waker)
+		switch ready {
+		case true:
+			if err != nil {
+				return err
+			}
+			if nc != nil {
+				co := newCommunicationOut(CommunicationOutCommit[Hash, Number, Signature, ID]{nc.Number, nc.Commit})
+				v.globalOut.Push(co)
+			} else {
+				break pastRounds
+			}
+		case false:
+			break pastRounds
+		}
+	}
+
+finalizedNotifications:
+	for {
+		select {
+		case notif := <-v.finalizedNotifications:
+			fHash := notif.Hash
+			fNum := notif.Number
+			round := notif.Round
+			commit := notif.Commit
+
+			v.inner.pastRounds.UpdateFinalized(fNum)
+			if v.setLastFinalizedNumber(fNum) {
+				err := v.env.FinalizeBlock(fHash, fNum, round, commit)
+				if err != nil {
+					return err
+				}
+			}
+
+			if fNum > v.lastFinalizedInRounds.Number {
+				v.lastFinalizedInRounds = HashNumber[Hash, Number]{fHash, fNum}
+			}
+		default:
+			// TODO: wake when messages come back
+			break finalizedNotifications
+		}
+	}
+
+	return nil
+}
+
+// Process all incoming messages from other nodes.
+//
+// Commit messages are handled with extra care. If a commit message references
+// a currently backgrounded round, we send it to that round so that when we commit
+// on that round, our commit message will be informed by those that we've seen.
+//
+// Otherwise, we will simply handle the commit and issue a finalisation command
+// to the environment.
+func (v *Voter[Hash, Number, Signature, ID]) processIncoming() error {
+	// TODO: implement waker chans
```

Comment:
is this needed now or something to do later?

Reply:
addressed